R/duplicate_argument_linter.R

```diff
@@ -0,0 +1,52 @@
+#' @describeIn linters checks for duplicate arguments in function calls.
+#' @param except a character vector of function names as exceptions.
+#' @export
+duplicate_argument_linter <- function(except = character()) {
+  Linter(function(source_file) {
+
+    if (is.null(source_file$full_xml_parsed_content)) return(list())
+
+    xml <- source_file$full_xml_parsed_content
+
+    xpath <- "//expr[EQ_SUB]"
+
+    calls <- xml2::xml_find_all(xml, xpath)
+
+    if (length(except)) {
+      calls_names <- xml2::xml_find_first(calls, "expr[1][count(*)=1]/*[1]")
+      calls_text <- vapply(
+        parse(text = xml2::xml_text(calls_names)),
+        as.character, character(1L)
+      )
+      calls <- calls[!(calls_text %in% except)]
+    }
+
+    result <- lapply(calls, function(call) {
+      args <- xml2::xml_find_all(call, "EQ_SUB/preceding-sibling::*[1]")
+      args_text <- vapply(
+        parse(text = xml2::xml_text(args), keep.source = FALSE),
+        as.character, character(1L)
+      )
+      is_dup_arg <- duplicated(args_text)
+      dup_args <- args[is_dup_arg]
+      if (length(dup_args)) {
+        line1 <- as.integer(xml2::xml_attr(dup_args, "line1"))
+        col1 <- as.integer(xml2::xml_attr(dup_args, "col1"))
+        col2 <- as.integer(xml2::xml_attr(dup_args, "col2"))
```

Comment:
Technically,

    tmp <- function(`
    ` = 42) {
    `
    `
    }

is valid R code, so `line2` is not guaranteed to be `line1` for call names.

Reply:
Here `line1` is only used for reporting. Do you have any suggestion what to do with `line2`? Does lintr support multi-lined lint result?

Reply:
The actual range is from `line1, col1` to `line2, col2`.

Reply:
IIRC we have some places where `col2` is automatically set to the end of the line if `line2 > line1`. I think that would be fine here, too. It's just a crude edge case anyway as noone in their right mind would use newlines in argument names other than for demonstrational purposes.

Reply:
`col2` tweaked to line ends. Do we need test case for such "crude edge case"?

Reply:
Sounds like a good idea to me.

Reply:
A crude edge test cases added.